[tf.data] add AutotuneOptions for performance tuning of dataset operations

kvignesh1420 opened the pull request:
This PR adds the `tf.data.AutotuneOptions` class, which is a split of options from `tf.data.experimental.OptimizationOptions`.

- [x] add new proto message for `AutotuneOptions` and extend the `Options` message.
- [x] add `tf.data.AutotuneOptions` class.
- [x] add `autotune` attribute to `Options`.
- [x] regenerate golden APIs.
- [x] handle backward-comaptibility with `options.experimental_optimization` attributes.
- [x] add/modify `options_test` w.r.t the new `autotune` attribute.
- [x] switch to `autotune_options` in:
    - [x] `tensorflow/core/kernels/data/finalize_dataset_op.cc`
    - [x] `tensorflow/core/data/dataset_utils.cc`
    - [x] `tensorflow/core/data/root_dataset.cc `
- [x] update RELEASE.md.

TEST LOG
```
IINFO: Elapsed time: 42.805s, Critical Path: 41.39s
INFO: 354 processes: 130 internal, 224 local.
INFO: Build completed successfully, 354 total actions
//tensorflow/python/data/kernel_tests:options_test                       PASSED in 3.6s

INFO: Build completed successfully, 354 total actions

---------------

INFO: Found 13 test targets...
INFO: Elapsed time: 23.775s, Critical Path: 22.98s
INFO: 44 processes: 31 internal, 13 local.
INFO: Build completed successfully, 44 total actions
//tensorflow/python/data/experimental/kernel_tests/optimization:map_parallelization_test (cached) PASSED in 3.8s
//tensorflow/python/data/experimental/kernel_tests/optimization:autotune_buffer_sizes_test PASSED in 5.0s
//tensorflow/python/data/experimental/kernel_tests/optimization:filter_fusion_test PASSED in 13.6s
//tensorflow/python/data/experimental/kernel_tests/optimization:grappler_test PASSED in 5.6s
//tensorflow/python/data/experimental/kernel_tests/optimization:grappler_test_gpu PASSED in 5.7s
//tensorflow/python/data/experimental/kernel_tests/optimization:hoist_random_uniform_test PASSED in 6.1s
//tensorflow/python/data/experimental/kernel_tests/optimization:map_and_batch_fusion_test PASSED in 5.6s
//tensorflow/python/data/experimental/kernel_tests/optimization:map_and_filter_fusion_test PASSED in 9.0s
//tensorflow/python/data/experimental/kernel_tests/optimization:map_fusion_test PASSED in 9.8s
//tensorflow/python/data/experimental/kernel_tests/optimization:noop_elimination_test PASSED in 6.6s
//tensorflow/python/data/experimental/kernel_tests/optimization:reorder_data_discarding_ops_test PASSED in 5.5s
//tensorflow/python/data/experimental/kernel_tests/optimization:shuffle_and_repeat_fusion_test PASSED in 5.5s
//tensorflow/python/data/experimental/kernel_tests/optimization:optimization_test PASSED in 23.0s
  Stats over 2 runs: max = 23.0s, min = 7.0s, avg = 15.0s, dev = 8.0s

```

cc: @jsimsa Let me know if I have missed something. Thanks!

kvignesh1420 committed:
[tf.data] add new class and attribute for AutotuneOptions

add AutotuneOptions to __init__.py

add autotune_options dep to dataset_ops target

regenerate golden APIs for AutotuneOptions

fix AutotuneOptions protobuf instance call

kvignesh1420 committed:
[tf.data] add autotune_options based options_tests

kvignesh1420 committed:
update RELEASE.md

kvignesh1420 committed:
[tf.data] switch from optimization_options to autotune_options for autotuning

jsimsa commented:
Thank you! Would it be possible for this CL to avoid the redundancy between `experimental_optimization.autotune*` options and `autotune.*` options? Could we remove the former and add a mechanism that would reroute getting and setting of the deprecated attributes to the new attributes? (similar to my changes to ThreadingOptions)

kvignesh1420 committed:
fix failing test case in dataset_utils_test

fix failing test case in finalize_dataset_op_test

fix autotune based optimization test failures

fix failing autotune_buffer_sizes_test

kvignesh1420 committed:
[tf.data] ensure optimization and autotune attr backward compatibility

[tf.data] refactor optimization and autotune attr handling

[tf.data] add deprecation warnings for autotune based optimization attrs

fix docstring of _get_optimization_autotune_opt_map

kvignesh1420 commented:
@jsimsa made the changes. Please let me know how it looks.

jsimsa commented on tensorflow/core/framework/dataset_options.proto:
```diff
@@ -139,6 +139,30 @@ enum ExternalStatePolicy {
   POLICY_FAIL = 2;
 }
 
+message AutotuneOptions {
```
We should also remove the corresponding fields from the OptimizationOptions and update the C++ to be backwards compatible (it should check whether the old OptimizationOptions fields exist and if so use them and if they don't use the new AutotuneOptions).

kvignesh1420 replied:
I have removed the redundant properties from `OptimizationOptions` proto message and retained the encoding code for the remaining properties.

kvignesh1420 replied:
Also since these proto message properties have been removed, the respective helper methods will also be removed. Thus, I have made the switch to autotune options directly.

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -3675,19 +3676,65 @@ class Options(options_lib.OptionsBase):
                       "Use options.threading instead.")
       return getattr(self, "threading")
+    elif name == "experimental_optimization":
+      value = getattr(self, "experimental_optimization")
+      autotune_opt = getattr(self, "autotune")
+      attr_map = self._get_optimization_autotune_opt_map()
+      for old_attr, new_attr in attr_map.items():
+        logging.warning("options.experimental_optimization.{}"
+        " is deprecated. Use options.autotune.{} instead".format(
+          old_attr, new_attr))
+        res = getattr(autotune_opt, new_attr)
+        if res:
+          setattr(value, old_attr, res)
+      return value
```
For similar reasons that prompted my changes to ThreadingOptions, we do not want to keep both the old and the new attributes and try to keep them in sync. Instead, we should remove the old attributes and rerouted getting and setting of the old attributes to the new attributes. 

You can accomplish this by having AutotuneOptions be a private attribute of the OptimizationOptions and have all accesses of `autotune` be routed through `OptimizationOptions` which either returns its `AutotuneOptions` instance (if user is accessing `autotune`) or updates the `AutotuneOptions` instance, if user is trying to update any of the removed attributes.

kvignesh1420 replied:
addressed in the commit.

kvignesh1420 committed:
[tf.data] refactor experimental_optimization with _autotune_option

fix private method access bug

handle recursive setattr

remove return from __setattr__

remove the autotune attr from Options

lint fixes

kvignesh1420 committed:
remove redundant options from OptimizationOptions proto

kvignesh1420 committed:
regenerate golden apis

kvignesh1420 commented:
@jsimsa can you please take a look.

jsimsa commented on tensorflow/core/data/dataset_utils.cc:
```diff
@@ -219,17 +219,18 @@ void DefaultOptimizationGraphRewrites(
   }
-  const bool has_autotune = optimization_options.optional_autotune_case() ==
-                            OptimizationOptions::kAutotune;
+  const auto& autotune_options = options.autotune_options();
```
the code will have to assume that the `options` object has been generated by code before your PR and so you will need to keep the old logic + add the new logic for backwards compatibility

this comment applies to all places that are accessing the `Options` object that was passed into a tf.data op

kvignesh1420 replied:
okay, will modify

jsimsa commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -165,18 +134,46 @@ class OptimizationOptions(options.OptionsBase):
+    }
+
+  def __getattr__(self, name):
+    # handle backward compatibility with deprecated options
```
This should log a deprecation warning similar to what we do for ThreadingOptions.

jsimsa commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -165,18 +134,46 @@ class OptimizationOptions(options.OptionsBase):
+      return getattr(self, name)
+
+  def __setattr__(self, name, value):
+    # handle backward compatibility with deprecated options
```
This should log a deprecation warning similar to what we do for ThreadingOptions.

jsimsa commented on tensorflow/python/data/ops/dataset_ops.py:
```diff
@@ -3676,18 +3676,25 @@ class Options(options_lib.OptionsBase):
       default_factory=threading_options.ThreadingOptions)
 
   def __getattr__(self, name):
+    # handle backward compatibility with deprecated options
```
you will still want to add `autotune` option attribute to `Options` so that it shows up in the API documentation ... (it will not be using by the implementation but that's an implementation detail)

kvignesh1420 replied:
@jsimsa if we add the `autotune` option to `Options` then the `__getattr__()` function will not be called when we are trying to access `options.autotune`.

jsimsa commented on tensorflow/python/data/experimental/ops/optimization_options.py:
```diff
@@ -165,18 +134,46 @@ class OptimizationOptions(options.OptionsBase):
       docstring="Whether to fuse shuffle and repeat transformations. If None, "
       "defaults to True.")
 
+  _autotune_option = options.create_option(
```
You should at a comment here that explain why do we do this.

kvignesh1420 committed:
[tf.data] handle optimization_options backward compatibility in .cc ops

kvignesh1420 commented on tensorflow/core/kernels/data/finalize_dataset_op.cc:
```diff
@@ -38,11 +38,26 @@ void GetModelDatasetParams(const Options& options,
+  // check for deprecated autotune options in optimization_options
+  // the condition below checks for the presence of the deprecated option
+  // and decides whether to go with the deprecated options or not.
+  try{
+    if(options.optimization_options().optional_autotune_buffers_case() ==
+      OptimizationOptions::kAutotuneBuffers){
```
@jsimsa I am trying to check if the deprecated attrs are present in the `optimization_options` or not. However, I am unable to catch this error:

```console
tensorflow/core/kernels/data/finalize_dataset_op.cc:47:28: error: 'kAutotuneBuffers' is not a member of 'tensorflow::data::OptimizationOptions'
       OptimizationOptions::kAutotuneBuffers){
```

What is the recommended approach to handle such exceptions/errors?

jsimsa commented:
@kvignesh1420 I have limited bandwidth to help you work on this PR, so let's put this on hold for now ... it is okay if this does not make it into the TF 2.6 release

kvignesh1420 commented:
@jsimsa apologies for the continuous requests. I wanted to finish this before the release so was a bit hasty. Thanks again!

kvignesh1420 committed:
try deprecated option

kvignesh1420 committed:
fix invalid option access

kvignesh1420 committed:
fix option access

gbaned commented:
@kvignesh1420  Can you please resolve conflicts? Thanks!

jsimsa commented:
My suggestion would be to close this PR. I plan to work on this myself later this quarter.

kvignesh1420 commented:
@jsimsa okay!
